### PR TITLE
[FLINK-26546] Extract Observer Interface

### PR DESCRIPTION
**Pull request by Aitozi**
This PR is mean to extract the `Observer` interface to simplify the observe logic. The main logic is aligned to the `Reconciler` interface

### CODE REVIEW

**Comment by Aitozi** [comment]
cc @morhidi @gyfora @wangyang0918

**Comment by gyfora** [comment]
Thanks @Aitozi, looks good at a first glance, I will review this in detail tomorrow morning :)

**Comment by wangyang0918** [comment]
Just a personal taste, compared to inheritance, I prefer composition. The problem of inheritance is that we do not know which methods in `BaseReconciler` and `BaseObserver` will be overridden by a subclass. Currently, it is not a major problem since we have very few methods in the interface.

Note: I do not mean we have to change in this PR.

**Comment by Aitozi** [comment]
@wangyang0918 Thanks for sharing your insights. IMO, one of the benefits of inherited `BaseXX` class is that it can provide the `template` ability and leave it implemented by the subclass. The composition seems can not achieve the same effect. 
For example, in the `BaseReconciler#cleanup` provide the common template part, and the both subclass only have to handle how to `shutdown` the corresponding component.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/BaseObserver.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+
+/** The base observer. */
+public abstract class BaseObserver implements Observer {
+
+    protected final Logger logger = LoggerFactory.getLogger(this.getClass());
```
Comment (bgeng777) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/BaseObserver.java:42]: nit: the logger can be static to reduce extra memory footprint.

Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/BaseObserver.java:42]: It's intended for the sharing of `logger` in subclass

**Comment by gyfora** [comment]
I also like the BaseXX approach, I think now we have a nice clean separation of observers, and reconcilers :)

**Pull request merged**